art/attacks/evasion/imperceptible_asr/imperceptible_asr.py

```diff
@@ -0,0 +1,621 @@
+        :return: Possible PSD maskers and indices.
+        """
+        # identify maskers. For simplification it is assumed that all maskers are tonal (vs. nontonal).
+        masker_idx = ss.argrelmax(psd_vector)[0]
+
+        # smooth maskers with their direct neighbors
+        psd_maskers = 10 * np.log10(np.sum([10 ** (psd_vector[masker_idx + i] / 10) for i in range(-1, 2)], axis=0))
+        return psd_maskers, masker_idx
+
+    def filter_maskers(self, maskers: np.ndarray, masker_idx: np.ndarray) -> Tuple[np.ndarray, np.ndarray]:
+        """Filter maskers.
+
+        First, discard all maskers that are below the absolute threshold of hearing. Second, reduce pairs of maskers
+        that are within 0.5 bark distance of each other by keeping the larger masker.
+
+        :param maskers: Masker PSD values.
+        :param masker_idx: Masker indices.
+        :return: Filtered PSD maskers and indices.
+        """
+        # filter on the absolute threshold of hearing
+        # note: deviates from Qin et al. implementation by filtering first on ATH and only then on bark distance
+        ath_condition = maskers > self.absolute_threshold_hearing[masker_idx]
+        masker_idx = masker_idx[ath_condition]
+        maskers = maskers[ath_condition]
+
+        # filter on the bark distance
+        bark_condition = np.ones(masker_idx.shape, dtype=bool)
+        i_prev = 0
+        for i in range(1, len(masker_idx)):
+            # find pairs of maskers that are within 0.5 bark distance of each other
+            if self.bark[i] - self.bark[i_prev] < 0.5:
+                # discard the smaller masker
+                i_todelete, i_prev = (i_prev, i_prev + 1) if maskers[i_prev] < maskers[i] else (i, i_prev)
+                bark_condition[i_todelete] = False
+            else:
+                i_prev = i
+        masker_idx = masker_idx[bark_condition]
+        maskers = maskers[bark_condition]
+
+        return maskers, masker_idx
+
+    def calculate_individual_threshold(self, maskers: np.ndarray, masker_idx: np.ndarray) -> np.ndarray:
+        """Calculate individual masking threshold with frequency denoted at bark scale.
+
+        :param maskers: Masker PSD values.
+        :param masker_idx: Masker indices.
+        :return: Individual threshold vector of shape `(window_size // 2 + 1)`.
+        """
+        delta_shift = -6.025 - 0.275 * self.bark
+        threshold = np.zeros(masker_idx.shape + self.bark.shape)
+        # TODO reduce for loop
+        for k, (masker_j, masker) in enumerate(zip(masker_idx, maskers)):
+            # critical band rate of the masker
+            z_j = self.bark[masker_j]
+            # distance maskees to masker in bark
+            delta_z = self.bark - z_j
+            # define two-slope spread function:
+            #   if delta_z <= 0, spread_function = 27*delta_z
+            #   if delta_z > 0, spread_function = [-27+0.37*max(PSD_masker-40,0]*delta_z
```

Comment:
Is this code still required?

Reply:
This is actually a comment that writes out the definition of the two-slope spread function :smile: